File: src/crypto/btcbech32.ts
/**
 * Bitcoin Bech32/Bech32m address.
 *
 * Validates mainnet SegWit addresses with the
 * Bech32 checksum for v0 and Bech32m for v1+.
 */

import type {
  GlobalValidator,
  ValidateResult,
} from "../types";

import { clean } from "#util/clean";
import { err } from "#util/result";

const CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
const GENERATORS = [
  0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd,
  0x2a1462b3,
] as const;
const BECH32_CONST = 1;
const BECH32M_CONST = 0x2bc830a3;
const CHECKSUM_LENGTH = 6;
const MIN_DATA_LENGTH = 11;
const MAX_DATA_LENGTH = 71;

const compact = (value: string): string =>
  clean(value, " ").toLowerCase();

const polymod = (values: readonly number[]): number => {
  let chk = 1;
  for (const value of values) {
    const top = chk >>> 25;
    chk = (((chk & 0x1ffffff) << 5) ^ value) >>> 0;
    for (let i = 0; i < GENERATORS.length; i += 1) {
      const generator = GENERATORS[i] ?? 0;
      if (((top >>> i) & 1) === 1) {
        chk = (chk ^ generator) >>> 0;
      }
    }
  }
  return chk >>> 0;
};

const hrpExpand = (hrp: string): number[] => {
  const expanded: number[] = [];
  for (const ch of hrp) {
    expanded.push(ch.charCodeAt(0) >>> 5);
  }
  expanded.push(0);
  for (const ch of hrp) {
    expanded.push(ch.charCodeAt(0) & 31);
  }
  return expanded;
};

const dataValues = (data: string): number[] | null => {
  const values: number[] = [];
  for (const ch of data) {
    const value = CHARSET.indexOf(ch);
    if (value === -1) return null;
    values.push(value);
  }
  return values;
};

const convertBits = (
  values: readonly number[],
  fromBits: number,
  toBits: number,
): number[] | null => {
  let accumulator = 0;
  let bits = 0;
  const maxValue = (1 << toBits) - 1;
  const result: number[] = [];

  for (const value of values) {
    if (value < 0 || value >>> fromBits !== 0) return null;
    accumulator = (accumulator << fromBits) | value;
    bits += fromBits;
    while (bits >= toBits) {
      bits -= toBits;
      result.push((accumulator >>> bits) & maxValue);
    }
  }

  if (bits >= fromBits) return null;
  if (((accumulator << (toBits - bits)) & maxValue) !== 0) {
    return null;
  }

  return result;
};

type Bech32Validation =
  | { valid: true }
  | {
      valid: false;
      code: "format" | "checksum" | "component";
    };

const validateBech32 = (
  rawValue: string,
): Bech32Validation => {
  const hasLower = /[a-z]/.test(rawValue);
  const hasUpper = /[A-Z]/.test(rawValue);
  if (hasLower && hasUpper)
    return { valid: false, code: "format" };

  const v = rawValue.toLowerCase();
  if (!v.startsWith("bc1"))
    return { valid: false, code: "component" };

  const data = v.slice(3);
  if (
    data.length < MIN_DATA_LENGTH ||
    data.length > MAX_DATA_LENGTH
  ) {
    return { valid: false, code: "format" };
  }

  const values = dataValues(data);
  if (values === null)
    return { valid: false, code: "format" };

  const expanded = hrpExpand("bc");
  for (const value of values) {
    expanded.push(value);
  }
  const check = polymod(expanded);

  const version = values.at(0);
  if (version === undefined || version > 16) {
    return { valid: false, code: "component" };
  }

  if (version === 0 && check !== BECH32_CONST) {
    return { valid: false, code: "checksum" };
  }
  if (version > 0 && check !== BECH32M_CONST) {
    return { valid: false, code: "checksum" };
  }

  const programValues = values.slice(1, -CHECKSUM_LENGTH);
  const program = convertBits(programValues, 5, 8);
  if (
    program === null ||
    program.length < 2 ||
    program.length > 40
  ) {
    return { valid: false, code: "component" };
  }

  if (
    version === 0 &&
    program.length !== 20 &&
    program.length !== 32
  ) {
    return { valid: false, code: "component" };
  }

  return { valid: true };
};

const validate = (value: string): ValidateResult => {
  const v = compact(value);
  const result = validateBech32(clean(value, " "));
  if (result.valid) return { valid: true, compact: v };

  if (result.code === "component") {
    return err(
      "INVALID_COMPONENT",
      "Bitcoin Bech32 address has an unsupported component",
    );
  }
  if (result.code === "checksum") {
    return err(
      "INVALID_CHECKSUM",
      "Bitcoin Bech32 address fails checksum validation",
    );
  }
  return err(
    "INVALID_FORMAT",
    "Bitcoin Bech32 address has an invalid format",
  );
};

const format = (value: string): string => compact(value);

const btcBech32: GlobalValidator = {
  scope: "global",
  name: "Bitcoin Bech32 Address",
  localName: "Bitcoin Bech32 Address",
  abbreviation: "BTC",
  aliases: [
    "Bitcoin address",
    "Bitcoin SegWit address",
    "Bech32 address",
    "Bech32m address",
  ] as const,
  candidatePattern:
    "(?:bc1[ac-hj-np-z02-9]{11,71}|BC1[AC-HJ-NP-Z02-9]{11,71})",
  entityType: "any",
  examples: [
    "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4",
    "bc1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3qccfmv3",
  ] as const,
  compact,
  format,
  validate,
};

export default btcBech32;
export {
  compact,
  convertBits,
  format,
  polymod,
  validate,
  validateBech32,
};

File: crates/stdnum-core/src/validators/crypto/btcbech32.rs
use crate::{
  BECH32_CHARSET, BECH32_CHECKSUM_LENGTH, BECH32_CONST, BECH32_GENERATORS,
  BECH32_MAX_DATA_LENGTH, BECH32_MIN_DATA_LENGTH, BECH32M_CONST,
  compact_without,
  types::{
    EntityType, ValidationError, ValidationResult, Validator, ValidatorScope,
    ValidatorSpec,
  },
};

const HRP_EXPANSION: &[u32] = &[3, 3, 0, 2, 3];

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Bech32ErrorCode {
  Format,
  Checksum,
  Component,
}

pub static VALIDATOR: Validator = Validator::new(ValidatorSpec {
  id: "crypto.btcbech32",
  name: "Bitcoin Bech32 Address",
  local_name: "Bitcoin Bech32 Address",
  abbreviation: "BTC",
  aliases: &[
    "Bitcoin address",
    "Bitcoin SegWit address",
    "Bech32 address",
    "Bech32m address",
  ],
  candidate_pattern: r"(?:bc1[ac-hj-np-z02-9]{11,71}|BC1[AC-HJ-NP-Z02-9]{11,71})",
  scope: ValidatorScope::Global,
  entity_type: EntityType::Any,
  source_url: None,
  lengths: &[],
  examples: &[
    "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4",
    "bc1qrp33g0q5c5txsp9arysrx4k6zdkfs4nce4xj0gdcccefvpysxf3qccfmv3",
  ],
  compact,
  format,
  validate,
  generate: None,
  parse: None,
});

#[must_use]
pub fn compact(value: &str) -> String {
  compact_without(value, &[' ']).to_lowercase()
}

#[must_use]
pub fn format(value: &str) -> String {
  compact(value)
}

#[must_use]
pub fn polymod(values: &[u32]) -> u32 {
  let mut check = 1_u32;
  for value in values {
    let top = check >> 25;
    check = ((check & 0x01ff_ffff) << 5) ^ value;
    for (index, generator) in BECH32_GENERATORS.iter().enumerate() {
      let shift = u32::try_from(index).unwrap_or(0);
      if ((top >> shift) & 1) == 1 {
        check ^= generator;
      }
    }
  }
  check
}

#[must_use]
pub fn convert_bits(
  values: &[u32],
  from_bits: u32,
  to_bits: u32,
) -> Option<Vec<u32>> {
  if from_bits >= u32::BITS || to_bits >= u32::BITS {
    return None;
  }
  let mut accumulator = 0_u32;
  let mut bits = 0_u32;
  let max_value = (1_u32 << to_bits).saturating_sub(1);
  let mut result = Vec::new();
  for value in values {
    if *value >> from_bits != 0 {
      return None;
    }
    accumulator = (accumulator << from_bits) | value;
    bits = bits.saturating_add(from_bits);
    while bits >= to_bits {
      bits = bits.saturating_sub(to_bits);
      result.push((accumulator >> bits) & max_value);
    }
  }
  if bits >= from_bits
    || ((accumulator << (to_bits.saturating_sub(bits))) & max_value) != 0
  {
    return None;
  }
  Some(result)
}

pub fn validate_bech32(raw_value: &str) -> Result<(), Bech32ErrorCode> {
  let has_lower = raw_value.chars().any(|ch| ch.is_ascii_lowercase());
  let has_upper = raw_value.chars().any(|ch| ch.is_ascii_uppercase());
  if has_lower && has_upper {
    return Err(Bech32ErrorCode::Format);
  }
  let value = raw_value.to_lowercase();
  let Some(data) = value.strip_prefix("bc1") else {
    return Err(Bech32ErrorCode::Component);
  };
  if data.len() < BECH32_MIN_DATA_LENGTH || data.len() > BECH32_MAX_DATA_LENGTH
  {
    return Err(Bech32ErrorCode::Format);
  }
  let mut values = Vec::with_capacity(data.len());
  for ch in data.chars() {
    let Some(position) = BECH32_CHARSET.find(ch) else {
      return Err(Bech32ErrorCode::Format);
    };
    values.push(u32::try_from(position).unwrap_or(0));
  }
  let mut expanded = HRP_EXPANSION.to_vec();
  expanded.extend(values.iter().copied());
  let check = polymod(&expanded);
  let Some(version) = values.first().copied() else {
    return Err(Bech32ErrorCode::Component);
  };
  if version > 16 {
    return Err(Bech32ErrorCode::Component);
  }
  if (version == 0 && check != BECH32_CONST)
    || (version > 0 && check != BECH32M_CONST)
  {
    return Err(Bech32ErrorCode::Checksum);
  }
  let end = values.len().saturating_sub(BECH32_CHECKSUM_LENGTH);
  let Some(program_values) = values.get(1..end) else {
    return Err(Bech32ErrorCode::Component);
  };
  let Some(program) = convert_bits(program_values, 5, 8) else {
    return Err(Bech32ErrorCode::Component);
  };
  if !(2..=40).contains(&program.len())
    || (version == 0 && !matches!(program.len(), 20 | 32))
  {
    return Err(Bech32ErrorCode::Component);
  }
  Ok(())
}

pub fn validate(value: &str) -> ValidationResult {
  let compact = compact(value);
  let raw = compact_without(value, &[' ']);
  match validate_bech32(&raw) {
    Ok(()) => Ok(compact),
    Err(Bech32ErrorCode::Component) => Err(ValidationError::InvalidComponent(
      "Bitcoin Bech32 address has an unsupported component",
    )),
    Err(Bech32ErrorCode::Checksum) => Err(ValidationError::InvalidChecksum(
      "Bitcoin Bech32 address fails checksum validation",
    )),
    Err(Bech32ErrorCode::Format) => Err(ValidationError::InvalidFormat(
      "Bitcoin Bech32 address has an invalid format",
    )),
  }
}
